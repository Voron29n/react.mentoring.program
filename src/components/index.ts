export * from './common/Button';
export * from './common/Logo';

export * from './containers/Header';
export * from './containers/Lightbox';
export * from './containers/MovieForm';
export * from './containers/MovieForm/MovieFormView';
export * from './containers/MovieList';
export * from './containers/MovieList/MovieListView';

export * from './service/DeleteMovie';
export * from './service/EditMovie';
export * from './service/AddMovie';
export * from './service/SearchMovie';

export * from './single/Confirm';
export * from './single/Congratulations';
export * from './single/DropDown';
export * from './single/DropDown/DropDownItem';
export * from './single/Footer';
export * from './single/FilterBar';
export * from './single/GenreBar';
export * from './single/GenreBar/GenreBarItem';
export * from './single/GenreCheckmarks';
export * from './single/MovieDetail';
export * from './single/MovieItem/MovieView';
export * from './single/MovieItem';
export * from './single/MovieInput';
export * from './single/MovieInput/MovieInputView/index';
export * from './single/SearchBar';
export * from './single/SortBar';
export * from './single/Spinner';
export * from './single/Redirect';

export * from './error/ErrorBoundary';
export * from './error/FetchErrorMessage';
export * from './error/UnknownPage';
