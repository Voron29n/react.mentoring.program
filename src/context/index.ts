import {
  ILightboxContext,
  lightboxActions,
  LightboxActionTypes,
  LightboxContext,
  lightboxMovieReducer
} from './lightbox';

export {
  LightboxContext,
  ILightboxContext,
  lightboxMovieReducer,
  LightboxActionTypes,
  lightboxActions
};
