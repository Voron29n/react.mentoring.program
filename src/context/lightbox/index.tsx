import React, { createContext, useMemo, useReducer } from 'react';
import {
  AddMovie,
  DeleteMovie,
  ILightboxProps,
  LightboxSize
} from 'components';
import { EditMovie } from 'components/service/EditMovie';
import {
  actions,
  LightboxActionTypes as ActionTypes,
  LightboxMovieAction
} from './actions';

export const lightboxActions = actions;

export const LightboxActionTypes = ActionTypes;

type MovieState = {
  lightbox: ILightboxProps;
};

export type ILightboxContext = {
  state: MovieState;
  dispatch: React.Dispatch<LightboxMovieAction>;
};
export const lightboxMovieReducer = (
  state: MovieState,
  action: LightboxMovieAction
) => {
  console.log(action);
  switch (action.type) {
    case LightboxActionTypes.ADD_MOVIE_FORM:
      return {
        lightbox: {
          lightboxSize: LightboxSize.LARGE,
          children: <AddMovie />
        }
      };
    case LightboxActionTypes.DELETE_MOVIE:
      return {
        lightbox: {
          lightboxSize: LightboxSize.SMALL,
          children: <DeleteMovie deletedMovie={action.payload.movie} />
        }
      };
    case LightboxActionTypes.EDIT_MOVIE_FORM:
      return {
        lightbox: {
          lightboxSize: LightboxSize.LARGE,
          children: <EditMovie editedMovie={action.payload.movie} />
        }
      };
    case LightboxActionTypes.CLOSE_LIGHTBOX:
      return {
        lightbox: null
      };
    case LightboxActionTypes.CHANGE_LIGHTBOX_SIZE:
      return {
        lightbox: {
          ...state.lightbox,
          lightboxSize: action.payload.lightboxSize
        }
      };
    default:
      return state;
  }
};

export const LightboxContext = createContext<ILightboxContext>(
  {} as ILightboxContext
);

const initialState = {
  lightbox: null
} as MovieState;

interface ILightboxContextProvider {
  children: JSX.Element | JSX.Element[];
}

export const LightboxContextProvider = ({
  children
}: ILightboxContextProvider) => {
  const [state, dispatch] = useReducer(lightboxMovieReducer, initialState);
  const lightboxValue = useMemo(() => ({ state, dispatch }), [state]);
  return (
    <LightboxContext.Provider value={lightboxValue}>
      {children}
    </LightboxContext.Provider>
  );
};
