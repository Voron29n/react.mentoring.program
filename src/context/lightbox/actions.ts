import { LightboxSize, Movie } from 'components';

export type LightboxMovieAction = {
  type: LightboxActionTypes;
  payload?: {
    movie?: Movie;
    lightboxSize?: LightboxSize;
  };
};

export enum LightboxActionTypes {
  ADD_MOVIE_FORM = 'ADD_MOVIE_FORM',
  EDIT_MOVIE_FORM = 'EDIT_MOVIE_FORM',
  DELETE_MOVIE = 'DELETE_MOVIE',
  CLOSE_LIGHTBOX = 'CLOSE_LIGHTBOX',
  CHANGE_LIGHTBOX_SIZE = 'CHANGE_LIGHTBOX_SIZE'
}

const createAction = ({ type, payload }: LightboxMovieAction) => ({
  type,
  payload
});

const deleteMovie = (movie: Movie) =>
  createAction({
    type: LightboxActionTypes.DELETE_MOVIE,
    payload: {
      movie
    }
  });

const addMovie = () =>
  createAction({ type: LightboxActionTypes.ADD_MOVIE_FORM });

const editMovie = (movie: Movie) =>
  createAction({
    type: LightboxActionTypes.EDIT_MOVIE_FORM,
    payload: { movie }
  });

const closeLightbox = () =>
  createAction({
    type: LightboxActionTypes.CLOSE_LIGHTBOX
  });

const changeLightboxSize = (lightboxSize: LightboxSize) =>
  createAction({
    type: LightboxActionTypes.CHANGE_LIGHTBOX_SIZE,
    payload: { lightboxSize }
  });

export const actions = {
  deleteMovie,
  addMovie,
  editMovie,
  closeLightbox,
  changeLightboxSize
};
