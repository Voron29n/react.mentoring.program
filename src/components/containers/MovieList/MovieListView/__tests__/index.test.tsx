import React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { render, RenderResult, screen } from '@testing-library/react';
import { MovieListView } from 'components';
import { LightboxContextProvider } from 'context/lightbox';
import { testMovies } from './mocks';

describe('MovieListView component', () => {
  const mockStore = configureMockStore([thunk]);
  const setMovieDetailMock = jest.fn();
  let movieListViewRender: RenderResult;

  beforeEach(() => {
    jest.mock('hooks', () => ({
      ...(jest.requireActual('hooks') as any),
      useActions: {
        setMovieDetail: setMovieDetailMock
      }
    }));
    const store = mockStore({
      movieDetail: {
        movieDetail: null
      }
    });
    movieListViewRender = render(
      <MemoryRouter initialEntries={['']}>
        <ReduxProvider store={store}>
          <LightboxContextProvider>
            <MovieListView
              movieList={testMovies}
              isError={false}
              isLoading={false}
            />
          </LightboxContextProvider>
        </ReduxProvider>
      </MemoryRouter>
    );
  });

  it('renders', () => {
    const { container } = movieListViewRender;
    expect(screen.getByText(testMovies.length)).toBeInTheDocument();
    expect(container.querySelectorAll('.movie__list > div').length).toBe(
      testMovies.length
    );
  });

  it('MovieListView snapshot', () => {
    expect(movieListViewRender).toMatchSnapshot();
  });
});

describe('MovieListView component with error', () => {
  let movieListViewRender: RenderResult;

  beforeEach(() => {
    movieListViewRender = render(
      <MovieListView movieList={testMovies} isError={true} isLoading={false} />
    );
  });

  it('renders', () => {
    const { container } = movieListViewRender;
    expect(container.querySelector('.movie__count')).not.toBeInTheDocument();
    expect(container.querySelector('.movie__list')).not.toBeInTheDocument();
    expect(container.querySelector('.spinner-box')).not.toBeInTheDocument();
  });

  it('MovieListView snapshot', () => {
    expect(movieListViewRender).toMatchSnapshot();
  });
});

describe('MovieListView component with loading', () => {
  let movieListViewRender: RenderResult;

  beforeEach(() => {
    movieListViewRender = render(
      <MovieListView movieList={testMovies} isError={false} isLoading={true} />
    );
  });

  it('renders', () => {
    const { container } = movieListViewRender;
    expect(container.querySelector('.movie__count')).not.toBeInTheDocument();
    expect(container.querySelector('.movie__list')).not.toBeInTheDocument();
    expect(container.querySelector('.spinner-box')).toBeInTheDocument();
  });

  it('MovieListView snapshot', () => {
    expect(movieListViewRender).toMatchSnapshot();
  });
});
