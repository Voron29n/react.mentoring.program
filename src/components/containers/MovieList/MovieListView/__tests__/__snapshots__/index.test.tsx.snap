// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieListView component MovieListView snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="movie__view "
      >
        <div
          class="movie__count"
        >
          <span>
            2
          </span>
          movies found
        </div>
        <div
          class="movie__list"
        >
          <div
            class="movie__item__container "
          >
            <div
              class="item"
            >
              <img
                alt="testTitle"
                class="poster"
                src="testPosterPath"
              />
              <div
                class="info"
              >
                <div>
                  testTitle
                </div>
                <div
                  class="release"
                >
                  2018
                </div>
              </div>
              <div
                class="genre"
              >
                testGenre1, testGenre2
              </div>
            </div>
          </div>
          <div
            class="movie__item__container "
          >
            <div
              class="item"
            >
              <img
                alt="testTitle 1"
                class="poster"
                src="testPosterPath"
              />
              <div
                class="info"
              >
                <div>
                  testTitle 1
                </div>
                <div
                  class="release"
                >
                  2018
                </div>
              </div>
              <div
                class="genre"
              >
                testGenre1, testGenre2
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="movie__view "
    >
      <div
        class="movie__count"
      >
        <span>
          2
        </span>
        movies found
      </div>
      <div
        class="movie__list"
      >
        <div
          class="movie__item__container "
        >
          <div
            class="item"
          >
            <img
              alt="testTitle"
              class="poster"
              src="testPosterPath"
            />
            <div
              class="info"
            >
              <div>
                testTitle
              </div>
              <div
                class="release"
              >
                2018
              </div>
            </div>
            <div
              class="genre"
            >
              testGenre1, testGenre2
            </div>
          </div>
        </div>
        <div
          class="movie__item__container "
        >
          <div
            class="item"
          >
            <img
              alt="testTitle 1"
              class="poster"
              src="testPosterPath"
            />
            <div
              class="info"
            >
              <div>
                testTitle 1
              </div>
              <div
                class="release"
              >
                2018
              </div>
            </div>
            <div
              class="genre"
            >
              testGenre1, testGenre2
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MovieListView component with error MovieListView snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="movie__view "
      >
        <div
          style="display: flex; justify-content: center; align-items: center;"
        >
          <iframe
            allowfullscreen=""
            class="giphy-embed"
            frameborder="0"
            src="https://giphy.com/embed/8L0Pky6C83SzkzU55a"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="movie__view "
    >
      <div
        style="display: flex; justify-content: center; align-items: center;"
      >
        <iframe
          allowfullscreen=""
          class="giphy-embed"
          frameborder="0"
          src="https://giphy.com/embed/8L0Pky6C83SzkzU55a"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MovieListView component with loading MovieListView snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="movie__view movie__loading"
      >
        <div
          class="spinner-box"
        >
          <svg
            class="spinner"
            viewBox="0 0 50 50"
          >
            <circle
              class="path"
              cx="25"
              cy="25"
              fill="none"
              r="20"
              stroke-width="5"
            />
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="movie__view movie__loading"
    >
      <div
        class="spinner-box"
      >
        <svg
          class="spinner"
          viewBox="0 0 50 50"
        >
          <circle
            class="path"
            cx="25"
            cy="25"
            fill="none"
            r="20"
            stroke-width="5"
          />
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
